src/Computation/Expression.cs

```diff
@@ -2,308 +2,500 @@
 using System.Collections.Generic;
 using System.Linq;
 using System.Numerics;
+using Microsoft.Z3;
 using Symbolica.Collection;
+using Symbolica.Computation.Exceptions;
+using Symbolica.Computation.Values;
+using Symbolica.Computation.Values.Constants;
+using Symbolica.Computation.Values.Symbolics;
 using Symbolica.Expression;
 
 namespace Symbolica.Computation
 {
     internal sealed class Expression : IExpression
     {
         private readonly ICollectionFactory _collectionFactory;
-        private readonly Lazy<BigInteger> _constant;
-        private readonly IBool[]? _constraints;
+        private readonly IValue[]? _constraints;
```

Comment:
An optional array seems like it introduces an unnecessary degree of freedom to me, as an empty array already indicates the absence of any values. I checked the code out and made this non-nullable. It seems like it still works with this change, but obviously forces a few changes in places that call the constructor etc because they obviously now need to supply `Array.Empty<>`. Personally, with my F# hat on, I prefer to eliminate the nulls and be explicit with `Array.Empty` at the call sites than save on a few characters of code. It's possible I've overlooked something more fundamental behind the reason for making these values nullable here though.

Reply:
The nullable is gone now.